### PR TITLE
Check safeMethods

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/2117

### CODE REVIEW

**Comment by erikzhang** [comment]
Maybe we can move `safemethods` to abi.

```json
{
  "name": "symbol",
  "parameters": [],
  "returntype": "String",
  "offset": 0,
  "safe": true
}
```

**Comment by shargon** [comment]
> Maybe we can move safemethods to abi.

I will do that

**Comment by erikzhang** [comment]
Native contracts' abi should be updated too.

**Comment by erikzhang** [comment]
@neo-project/ngd-shanghai Can you test it?

**Comment by shargon** [comment]
@neo-project/ngd-shanghai you can use https://github.com/neo-project/neo-devpack-dotnet/pull/401

File: src/neo/SmartContract/ApplicationEngine.Contract.cs [hunk 1]
```diff
@@ -131,9 +131,16 @@ private void CallContractInternal(UInt160 contractHash, string method, Array arg
             ContractMethodDescriptor md = contract.Manifest.Abi.GetMethod(method);
             if (md is null) throw new InvalidOperationException($"Method {method} Does Not Exist In Contract {contractHash}");
 
-            ContractState currentContract = Snapshot.Contracts.TryGet(CurrentScriptHash);
-            if (currentContract?.CanCall(contract, method) == false)
-                throw new InvalidOperationException($"Cannot Call Method {method} Of Contract {contractHash} From Contract {CurrentScriptHash}");
+            if (md.Safe)
+            {
+                flags &= ~CallFlags.AllowModifyStates;
```
Comment (Tommo-L) [inline src/neo/SmartContract/ApplicationEngine.Contract.cs:136]: If a method is marked as `Safe`, will it not be allowed to modify states?

Comment (erikzhang) [reply src/neo/SmartContract/ApplicationEngine.Contract.cs:136]: Yes.

**Comment by fyrchik** [comment]
What about having `RequiredCallFlags` instead of `Safe`?
This will unify native and other contracts, while also providing some security.

**Comment by shargon** [comment]
> What about having RequiredCallFlags instead of Safe?

I think developers should know more about the core to use this properly, it is easier for adoption to make things easier.

**Review by erikzhang — DISMISSED** [review]

**Review by erikzhang — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Pull request merged**